File: 210302150_MujkanovićEnita_homework_1.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

public class QuizQuestion
{
    public string Question { get; set; }
    public List<string> Options { get; set; }
    public string Answer { get; set; }
    public int? SelectedOptionIndex { get; set; } //nullable int to represent if an option was selected
}

class Program
{
    static async Task Main()
    {
        string jsonUrl = "https://raw.githubusercontent.com/mujkenit/Small-Quiz-Application/main/pitanja.json";

        try
        {
            string jsonData = await DownloadJsonDataAsync(jsonUrl);

            if (!string.IsNullOrEmpty(jsonData))
            {
                Console.WriteLine("Press enter to start the Quiz!");
                Console.ReadLine();

                PrintRandomizedQuestionsWithAnswers(jsonData);
            }
            else
            {
                Console.WriteLine("Failed to download JSON data.");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"An error occurred: {ex.Message}");
        }
    }

    static async Task<string> DownloadJsonDataAsync(string url)
    {
        using (HttpClient httpClient = new HttpClient())
        {
            HttpResponseMessage response = await httpClient.GetAsync(url);

            if (response.IsSuccessStatusCode)
            {
                return await response.Content.ReadAsStringAsync();
            }

            return null;
        }
    }

    static void PrintRandomizedQuestionsWithAnswers(string jsonData)
    {
        //parses JSON data
        JObject json = JObject.Parse(jsonData);

        //accesses the "Quiz" property and the "questions" array
        JArray questionsArray = (JArray)json["Quiz"]["questions"];

        if (questionsArray != null)
        {
            //randomizes the order of questions using Fisher-Yates shuffle algorithm
            List<QuizQuestion> randomizedQuestions = new List<QuizQuestion>();

            foreach (var questionToken in questionsArray)
            {
                var question = questionToken.ToObject<QuizQuestion>();
                randomizedQuestions.Add(question);
            }

            Random rand = new Random();
            int n = randomizedQuestions.Count;

            while (n > 1)
            {
                n--;
                int k = rand.Next(n + 1);
                var value = randomizedQuestions[k];
                randomizedQuestions[k] = randomizedQuestions[n];
                randomizedQuestions[n] = value;
            }

            //flag to check if it's the first question
            bool firstQuestion = true;

            //variables to track correct and incorrect answers
            int correctAnswers = 0;
            int incorrectAnswers = 0;

            //iterates through each randomized question
            for (int i = 0; i < randomizedQuestions.Count; i++)
            {
                
                if (!firstQuestion)
                {
                    Console.WriteLine("Press enter to move on to the next question!");
                    Console.ReadLine();
                }
                else
                {
                    firstQuestion = false;
                }

                Console.WriteLine($"{i + 1}. {randomizedQuestions[i].Question}");

                //prints answers
                for (int j = 0; j < randomizedQuestions[i].Options.Count; j++)
                {
                    Console.WriteLine($"   {char.ConvertFromUtf32(97 + j)}. {randomizedQuestions[i].Options[j]}");
                }

                //forces user to input
                string userAnswer;
                do
                {
                    Console.Write("Your answer (a, b, c, d): ");
                    userAnswer = Console.ReadLine();

                    if (userAnswer == null || userAnswer.Length != 1 || !"abcd".Contains(userAnswer.ToLower()))
                    {
                        Console.WriteLine("Invalid answer. Please enter a valid option (a, b, c, d).");
                    }
                } while (userAnswer == null || userAnswer.Length != 1 || !"abcd".Contains(userAnswer.ToLower()));

                int selectedOptionIndex = userAnswer.ToLower()[0] - 'a';
                randomizedQuestions[i].SelectedOptionIndex = selectedOptionIndex;

                //checks if the user's answers are correct
                if (selectedOptionIndex == randomizedQuestions[i].Options.FindIndex(opt => opt.ToLower() == randomizedQuestions[i].Answer.ToLower()))
                {
                    Console.WriteLine("Correct!");
                    correctAnswers++;
                }
                else
                {
                    Console.WriteLine($"Incorrect. The correct answer is {randomizedQuestions[i].Answer.ToUpper()}.");
                    incorrectAnswers++;
                }

                
                Console.WriteLine();
            }

            //pritns the results
            Console.WriteLine($"You have finished this quiz :)");
            Console.WriteLine($"Number of Correct Answers: {correctAnswers}");
            Console.WriteLine($"Number of Incorrect Answers: {incorrectAnswers}");
        }
        else
        {
            Console.WriteLine("No questions found in the JSON data.");
        }
    }
}

File: 210302150_MujkanovićEnita_homework_1.py
#importing/reading a .json file into my Python code
import json
import random

def ask_question(question):
    print(question['question'])
    for i, option in enumerate(question['options'], start=1):
        print(f"{i}. {option}")
    
    while True:
        try:
            user_answer = int(input("Your answer (enter the number corresponding to your choice): "))
            if 1 <= user_answer <= len(question['options']):
                return user_answer
            else:
                print("Invalid choice. Please enter a number between 1 and", len(question['options']))
        except ValueError:
            print("Invalid input. Please enter a number.")


#open the file in read mode using a relative path

file_path = "pitanja.json"
with open(file_path, 'r') as file:
    data = json.load(file)  #loads the .json data from the file

#validated the structure of each qs
questions = data['Quiz']['questions'] #extracts the list of qs from the loaded data
for idx, question in enumerate(questions, start=1):
    if not all(key in question for key in ['question', 'options', 'answer']):
        raise ValueError(f"Invalid structure for question {idx}. Each question must have 'question', 'options', and 'answer' keys.")
    

# num_random_questions = 5 #num of random qs I want to select
num_questions_to_ask = len(questions)

random_questions = random.sample(questions, num_questions_to_ask) 
#randomly selects qs

#for idx, question in enumerate(random_questions, start=1):
#    print(f"\nQuestion {idx}: {question['question']}")
#    for i, option in enumerate(question['options'], start=1):
#        print(f"{i}. {option}")
#    print("")
#this one displays the randomly selected qs    

#quiz functionality
correct_answers = 0
for idx, question in enumerate(random_questions, start=1):
    input("Press enter to start the next question!")

    user_choice = ask_question(question)
    correct_choice = question['options'].index(question['answer']) + 1
    if user_choice == correct_choice:
        print("Correct!\n")
        correct_answers += 1
    else:
        print(f"Incorrect. The correct answer is {correct_choice}: {question['answer']}\n")

# Display quiz results
print(f"You answered {correct_answers} out of {num_questions_to_ask} questions correctly.")

